include/LLVMSPIRVOpts.h

```diff
@@ -99,12 +102,63 @@ class TranslatorOpts {
 
   void enableGenArgNameMD() { GenKernelArgNameMD = true; }
 
+  // Stores values for specialization constants. Potentially, can be used for
+  // implementation of clSetProgramSpecializationConstant API.
+  // Duplicate spec ids are not allowed, returns true in this case.
+  // Returns false as success.
+  // TODO: Should we provide an API which will check for CL_INVALID_SPEC_ID and
+  // CL_INVALID_VALUE erros?
+  bool addSpecConst(unsigned SpecId, size_t Size /*in bytes*/,
+                    const void *Value) {
+    if (std::find(ExternalSpecializationIds.begin(),
+                  ExternalSpecializationIds.end(),
+                  SpecId) != ExternalSpecializationIds.end()) {
+      return true;
+    }
+    ExternalSpecializationIds.push_back(SpecId);
+    ExternalSpecializationValues.emplace_back(static_cast<const char *>(Value),
```

Comment:
In `checkSpecializationConstant` you assume that `Size` is always 8, 16, 32 or 64, but you don't have any checks if this is not the case. I suggest to at least put some assertions here or into `switch (Buffer.size()) {` from `checkSpecializationConstant`

Reply:
Found the assert within SPIRVReader.cpp, but I still think that it is needed somewhere in this header as well, because users might initialize spec constants via `TranslatorOpts` in their apps